### PR TITLE
Fixes for lots of issues

### LINKED ISSUE #517: Support for HEX colour codes
# Bug Report / Suggestion?
With the release of Minecraft 1.16, there is an added support for HEX colour codes, and they work fine with the other plugins used on the server however not with NTE. I'm unaware as to whether NTE does not support this yet, or if it is a bug causing it to now work.

**Type /ver NametagEdit and post the output.**
4.4.10

**What version of Bukkit/Spigot/PaperSpigot are you using? Type /ver**
git-Paper-83 (MC: 1.16.1) (Implementing API version 1.16.1-R0.1-SNAPSHOT)

**What plugins are you using? Type /plugins**
N/A

**Please explain your issue. How do you trigger it?**
[https://prnt.sc/tuh44k](url)

**Are there any errors in the console? Please use: https://pastebin.com**
N/A

### LINKED ISSUE #560: WARN
[17:46:52 WARN]: [NametagEdit] Loaded class net.luckperms.api.LuckPermsProvider from LuckPerms v5.2.76 which is not a depend, softdepend or loadbefore of this plugin.

### LINKED ISSUE #577: Feature Request
Can you add an api method to hide player nametag **above head**?

### LINKED ISSUE #585: hexadecimal?
is it possible to use hexadecimal codes?

### LINKED ISSUE #589: Don't working command /nte reload
# Bug Report

**Type /ver NametagEdit and post the output.**
NametagEdit version 4.4.10

**What version of Bukkit/Spigot/PaperSpigot are you using? Type /ver**
This server is running Paper version git-Paper-638 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)
You are 4 version(s) behind
Previous version: git-Paper-362 (MC: 1.16.4)

**What plugins are you using? Type /plugins**
Plugins (42): AdvancedGUI*, ChestCommands*, Citizens, CMI, CMILib, CommandNPC*, Converter*, CoreProtect, CrazyCrates, CustomJoinItems*, DeluxeMenus, DiscordSRV, DKCoins*, EasyBackup, eZProtector, FastAsyncWorldEdit (WorldEdit), HeadDatabase*, HopeCommander-Spigot*, ImageOnMap, JukeBox, LiteJoin*, LuckPerms, McNative*, NametagEdit*, NoCheatPlus, NoteBlockAPI, PlaceholderAPI, ProtocolLib, Reqn*, Skript, skript-placeholders*, SkriptJSON*, skUtilities*, TitleManager, TpLogin*, Vault, VentureChat, ViaBackwards, ViaRewind, ViaVersion, WorldBorder, WorldGuard

**Please explain your issue. How do you trigger it?**
When you start the server or when you write the command /nte reload the following comes out:
`[19:03:33 WARN]: [NametagEdit] Plugin NametagEdit v4.4.10 generated an exception while executing task 153391
java.lang.RuntimeException: java.lang.ClassNotFoundException: com.mysql.jdbc.jdbc2.optional.MysqlDataSource
        at com.zaxxer.hikari.util.UtilityElf.createInstance(UtilityElf.java:102) ~[?:?]
        at com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:316) ~[?:?]
        at com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:109) ~[?:?]
        at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:108) ~[?:?]
        at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112) ~[?:?]
        at com.nametagedit.plugin.storage.database.tasks.DataDownloader.run(DataDownloader.java:40) ~[?:?]
        at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftTask.run(CraftTask.java:100) ~[patched_1.16.5.jar:git-Paper-638]
        at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54) ~[patched_1.16.5.jar:git-Paper-638]
        at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[patched_1.16.5.jar:git-Paper-638]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
        at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.lang.ClassNotFoundException: com.mysql.jdbc.jdbc2.optional.MysqlDataSource
        at java.net.URLClassLoader.findClass(Unknown Source) ~[?:?]
        at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:182) ~[patched_1.16.5.jar:git-Paper-638]
        at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:111) ~[patched_1.16.5.jar:git-Paper-638]
        at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:?]
        at java.lang.ClassLoader.loadClass(Unknown Source) ~[?:?]
        at com.zaxxer.hikari.util.UtilityElf.createInstance(UtilityElf.java:89) ~[?:?]
        ... 11 more`

**Are there any errors in the console? Please use: https://pastebin.com**
https://pastebin.com/pNQvfmdA

### LINKED ISSUE #593: MySQL error after started the server.
# Bug Report

**Type /ver NametagEdit and post the output.**
v4.4.16 - Latest stable dev version.

**What version of Bukkit/Spigot/PaperSpigot are you using? Type /ver**
aper version git-Paper-682 (MC: 1.16.5) (Implementing API version 1.16.5-R0.1-SNAPSHOT)

**What plugins are you using? Type /plugins**
Plugins (57): BarrierPlus, BetterBowTrails*, BetterDrops, BossShopPro, CMI, CMIEInjector*, CMILib, CommandConditions, CommandSync*, CorePlus, CoreProtect, DiscordSRV, Emotes, EntityPlus, FeatherBoard*, FriendsPremium, HeadDatabase, HotkeyPlus, IllegalStack, InventorySaver, ItemJoin, JetsRepairTokens, LibsDisguises, LiquidTanks, LiteBans, LotteryPlus, LuckPerms, LuckPermsMVdWHook*, MaSuiteCore, MaSuiteHomes, MaSuiteTeleports, MaSuiteWarps, MobFarmManager, MorphTool, Multiverse-Core, MVdWPlaceholderAPI*, MyCommand, MyPet, MysqlPlayerDataBridge, NametagEdit*, ParticleHats, PlaceholderAPI, PlayerdataPlus, ProtocolLib, PvPManager, RegionPlus, Residence, SelectionVisualizer, ServerPlus, SkinsRestorer, SlimeChunkPlus, SurvivalMechanics, TradeMe, Vault, Vehicles, WorldEdit, WorldGuard

**Please explain your issue. How do you trigger it?**
The error occurred after server started.
`java.lang.ClassNotFoundException: com.mysql.jdbc.jdbc2.optional.MysqlDataSource`

**Are there any errors in the console? Please use: https://pastebin.com**
Full console log.
https://pastebin.com/zzNsRB8e

```
[04:27:05] [Craft Scheduler Thread - 74/WARN]: [NametagEdit] Plugin NametagEdit v4.4.16 generated an exception while executing task 1698
java.lang.RuntimeException: java.lang.ClassNotFoundException: com.mysql.jdbc.jdbc2.optional.MysqlDataSource
	at com.zaxxer.hikari.util.UtilityElf.createInstance(UtilityElf.java:93) ~[?:?]
	at com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:317) ~[?:?]
	at com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:111) ~[?:?]
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:106) ~[?:?]
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:97) ~[?:?]
	at com.nametagedit.plugin.storage.database.tasks.DatabaseUpdater.run(DatabaseUpdater.java:26) ~[?:?]
	at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftTask.run(CraftTask.java:100) ~[patched_1.16.5.jar:git-Paper-682]
	at org.bukkit.craftbukkit.v1_16_R3.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54) ~[patched_1.16.5.jar:git-Paper-682]
	at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[patched_1.16.5.jar:git-Paper-682]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.ClassNotFoundException: com.mysql.jdbc.jdbc2.optional.MysqlDataSource
	at org.bukkit.plugin.java.PluginClassLoader.loadClass0(PluginClassLoader.java:153) ~[patched_1.16.5.jar:git-Paper-682]
	at org.bukkit.plugin.java.PluginClassLoader.loadClass(PluginClassLoader.java:114) ~[patched_1.16.5.jar:git-Paper-682]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:522) ~[?:?]
	at com.zaxxer.hikari.util.UtilityElf.createInstance(UtilityElf.java:80) ~[?:?]
	... 11 more
[04:27:05] [Server thread/WARN]: Warning: Nashorn engine is planned to be removed from a future JDK release
```

### LINKED ISSUE #606: Update API & Developers to last version
[Click here](https://github.com/sgtcaze/NametagEdit/blob/master/documentation/Developers.creole)

```
<dependency>
    <groupId>com.nametagedit</groupId>
    <artifactId>nametagedit</artifactId>
    <version>4.1.0</version>
</dependency>
```

to

```
<dependency>
    <groupId>com.nametagedit</groupId>
    <artifactId>nametagedit</artifactId>
    <version>4.4.16</version>
</dependency>
```

### LINKED ISSUE #607: 1.17 Support
Is there going to be support added for 1.17

### LINKED ISSUE #609: Driver does not support get/set network timeout for connections
Dev Build: 4.5.0 (dev2 branch)

```
INFO NametagEdit Pool - Driver does not support get/set network timeout for connections. (com.mysql.jdbc.JDBC4Connection.getNetworkTimeout()I
```

- MySQL still functional.

### LINKED ISSUE #610: Compatibilidad o Actualizacion a 1.17
# Bug Report

**Type /ver NametagEdit and post the output.**

**What version of Bukkit/Spigot/PaperSpigot are you using? Type /ver**
Paper 1.17

**What plugins are you using? Type /plugins**
-Essentials,Luckperms,Skin Restorer

**Please explain your issue. How do you trigger it?**
-Al configurarlo me sale error de comando interno

**Are there any errors in the console? Please use: https://pastebin.com**
-me sale error de comando interno

### PR DESCRIPTION
**Pull request by Lorenzo0111**
This pull request includes lots of fixes. I've merged [Aurelien30000/NametagEdit](https://github.com/Aurelien30000/NametagEdit) and some commits.

Closes #606, closes #593, closes #589, closes #585, closes #609, closes #607, closes #610 closes #560, closes #577 and closes #517

### CODE REVIEW

**Comment by JustOneMoreBlock** [comment]
@Lorenzo0111 Please use dev2.

**Comment by JustOneMoreBlock** [comment]
Also tested your build. Did you get Hex Codes working? They don't seem to be working here.

```
/nte player NametagEdit prefix '&#60F50BHi &r'
```
- If this worked. It would be green.
- If not perhaps would have been `&6` which would have been gold.
- It just applies the entire thing as text.

https://i.imgur.com/Uw7se6F.png

**Comment by Aurelien30000** [comment]
> Also tested your build. Did you get Hex Codes working? They don't seem to be working here.
> 
> ```
> /nte player NametagEdit prefix '&#60F50BHi &r'
> ```
> 
> * If this worked. It would be green.
> * If not perhaps would have been `&6` which would have been gold.
> * It just applies the entire thing as text.
> 
> https://i.imgur.com/Uw7se6F.png

You should take a look at this method https://github.com/Nathat23/StackMob-5/blob/master/src/main/java/uk/antiperson/stackmob/utils/Utilities.java#L36 and the related rgx pattern.

**Comment by Lorenzo0111** [comment]
I'm recreating it

**Comment by Lorenzo0111** [comment]
That's strange, If I print the colored string in the chat it works but on the nametag it gives problems..

**Review by Aurelien30000 — APPROVED** [review]

**Comment by Lorenzo0111** [comment]
Now also closes #577. 
Already tested

**Comment by Lorenzo0111** [comment]
If you need other help just assign an issue to me

File: src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java [hunk 1]
```diff
@@ -94,14 +94,14 @@ public PacketWrapper(String name, String prefix, String suffix, int param, Colle
...
-                        PacketAccessor.VISIBILITY.set(packet, "always");
+                        PacketAccessor.VISIBILITY.set(packet, visible ? "always" : "never");
                     }
                 } else {
                     // 1.17+
                     PacketAccessor.PACK_OPTION.set(packetParams, 1);
 
                     if (PacketAccessor.VISIBILITY != null) {
-                        PacketAccessor.VISIBILITY.set(packetParams, "always");
+                        PacketAccessor.VISIBILITY.set(packet, visible ? "always" : "never");
```
Comment (Aurelien30000) [inline src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java]: ``packetParams``, very important

**Review by Aurelien30000 — CHANGES_REQUESTED** [review]

File: src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java [hunk 2]
```diff
@@ -94,14 +94,14 @@ public PacketWrapper(String name, String prefix, String suffix, int param, Colle
                     PacketAccessor.PACK_OPTION.set(packet, 1);
 
                     if (PacketAccessor.VISIBILITY != null) {
-                        PacketAccessor.VISIBILITY.set(packet, visible ? "always" : "never");
+                        PacketAccessor.VISIBILITY.set(packetParams, visible ? "always" : "never");
```
Comment (Aurelien30000) [inline src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java]: Not again 😅 
Pre-1.17, ``packet``, post-1.17 ``packetParams``, sorry, wasn't clear.

**Review by Aurelien30000 — CHANGES_REQUESTED** [review]

File: src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java (hunk 2 above)
Comment (Lorenzo0111) [reply src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java]: Have I to put a check?

Comment (Aurelien30000) [reply src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java]: No, I already implemented the check, loot at ``// 1.17+`` comment. Just like that:

```java
                if (!PacketAccessor.isParamsVersion()) {
                    PacketAccessor.PACK_OPTION.set(packet, 1);

                    if (PacketAccessor.VISIBILITY != null) {
                        PacketAccessor.VISIBILITY.set(packet, visible ? "always" : "never");
                    }
                } else {
                    // 1.17+
                    PacketAccessor.PACK_OPTION.set(packetParams, 1);

                    if (PacketAccessor.VISIBILITY != null) {
                        PacketAccessor.VISIBILITY.set(packetParams, visible ? "always" : "never");
                    }
                }
```

Comment (Lorenzo0111) [reply src/main/java/com/nametagedit/plugin/packets/PacketWrapper.java]: oh lol, i didn't see it

**Review by Aurelien30000 — APPROVED** [review]

**Comment by JustOneMoreBlock** [comment]
[NTE Test](https://www.youtube.com/watch?v=d8SEa1YZ4ZA)

- Doesn't appear `HEX Codes` are still working and I compiled your version and tested it. @Lorenzo0111

**Comment by JustOneMoreBlock** [comment]
I tried switching to `MySQL`. Looks like you're still using your `cj` driver over the `jdbcUrl`

```
26.06 22:54:07 [Server] WARN NametagEdit Pool - using dataSourceClassName and ignoring jdbcUrl.
26.06 22:54:07 [Server] INFO NametagEdit Pool - Starting...
26.06 22:54:07 [Server] WARN Plugin NametagEdit v4.5.0 generated an exception while executing task 431
26.06 22:54:07 [Server] INFO java.lang.RuntimeException: java.lang.ClassCastException: Cannot cast com.mysql.cj.jdbc.Driver to javax.sql.DataSource
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.util.UtilityElf.createInstance(UtilityElf.java:107) ~[?:?]
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:327) ~[?:?]
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:114) ~[?:?]
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:108) ~[?:?]
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112) ~[?:?]
26.06 22:54:07 [Server] INFO at com.nametagedit.plugin.storage.database.tasks.DatabaseUpdater.run(DatabaseUpdater.java:26) ~[?:?]
26.06 22:54:07 [Server] INFO at org.bukkit.craftbukkit.v1_17_R1.scheduler.CraftTask.run(CraftTask.java:100) ~[patched_1.17.jar:git-Paper-"ce66102"]
26.06 22:54:07 [Server] INFO at org.bukkit.craftbukkit.v1_17_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:54) ~[patched_1.17.jar:git-Paper-"ce66102"]
26.06 22:54:07 [Server] INFO at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[patched_1.17.jar:git-Paper-"ce66102"]
26.06 22:54:07 [Server] INFO at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[?:?]
26.06 22:54:07 [Server] INFO at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[?:?]
26.06 22:54:07 [Server] INFO at java.lang.Thread.run(Thread.java:831) [?:?]
26.06 22:54:07 [Server] INFO Caused by: java.lang.ClassCastException: Cannot cast com.mysql.cj.jdbc.Driver to javax.sql.DataSource
26.06 22:54:07 [Server] INFO at java.lang.Class.cast(Class.java:3818) ~[?:?]
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.util.UtilityElf.createInstance(UtilityElf.java:96) ~[?:?]
26.06 22:54:07 [Server] INFO ... 11 more
26.06 22:54:07 [Server] WARN NametagEdit Pool - using dataSourceClassName and ignoring jdbcUrl.
26.06 22:54:07 [Server] INFO NametagEdit Pool - Starting...
26.06 22:54:07 [Server] WARN Plugin NametagEdit v4.5.0 generated an exception while executing task 435
26.06 22:54:07 [Server] INFO java.lang.RuntimeException: java.lang.ClassCastException: Cannot cast com.mysql.cj.jdbc.Driver to javax.sql.DataSource
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.util.UtilityElf.createInstance(UtilityElf.java:107) ~[?:?]
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:327) ~[?:?]
26.06 22:54:07 [Server] INFO at com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:114) ~[?:?]
26.06 22:54:10 [Multicraft] Skipped 13 lines due to rate limit (30/s)
```

**Comment by Lorenzo0111** [comment]
> [NTE Test](https://www.youtube.com/watch?v=d8SEa1YZ4ZA)
> 
> * Doesn't appear `HEX Codes` are still working and I compiled your version and tested it. @Lorenzo0111

I know, as I [said](https://github.com/sgtcaze/NametagEdit/pull/611#issuecomment-868975436) it is strange and it does not work.

**Review by Aurelien30000 — APPROVED** [review]

**Comment by JustOneMoreBlock** [comment]
My competitor has RGB support.

https://www.spigotmc.org/resources/tab-1-5-x-1-17-free-version.57806/
GitHub: https://github.com/NEZNAMY/TAB

I wouldn't necessarily use their code, however, may find what's not allowing NTE to work properly. Keep in mind, I've used code from other sources before, however, it specifically states I can and I referenced them. So, I like to give credit where it's due.

**Comment by Lorenzo0111** [comment]
> My competitor has RGB support.
> 
> https://www.spigotmc.org/resources/tab-1-5-x-1-17-free-version.57806/
> GitHub: https://github.com/NEZNAMY/TAB
> 
> I wouldn't necessarily use their code, however, may find what's not allowing NTE to work properly. Keep in mind, I've used code from other sources before, however, it specifically states I can and I referenced them. So, I like to give credit where it's due.

I have an idea, I’ll try in some hours

**Comment by Aurelien30000** [comment]
We should figure out the exact format required and used by vanilla minecraft/client. Perhaps that's something like ``§x§R§R§G§G§B§B`` <=> ``#RRGGBB``

**Comment by Lorenzo0111** [comment]
> 
> 
> > My competitor has RGB support.
> > https://www.spigotmc.org/resources/tab-1-5-x-1-17-free-version.57806/
> > GitHub: https://github.com/NEZNAMY/TAB
> > I wouldn't necessarily use their code, however, may find what's not allowing NTE to work properly. Keep in mind, I've used code from other sources before, however, it specifically states I can and I referenced them. So, I like to give credit where it's due.
> 
> I have an idea, I’ll try in some hours

It does not work, I've tried with bungee ChatColor#of but nothing

**Comment by Lorenzo0111** [comment]
> 
> 
> We should figure out the exact format required and used by vanilla minecraft/client. Perhaps that's something like `§x§R§R§G§G§B§B` <=> `#RRGGBB`

Same thing, if I print it in chat it works but not in the nametag

**Comment by Aurelien30000** [comment]
https://github.com/sgtcaze/NametagEdit/pull/615 Your formatter wasn't the cause of the issues!

**Comment by Lorenzo0111** [comment]
@JustOneMoreBlock now the hex translator works

**Comment by Lorenzo0111** [comment]
Thanks for merging 🙂.
Now you can close the issues that I’ve linked to this pull request

**Pull request merged**